examples/platforms/simulation/simul_utils.c

```diff
@@ -90,43 +206,156 @@ void utilsInitSocket(utilsSocket *aSocket, uint16_t aPortBase)
+        addr6 = NULL;
+    }
+    else if (addr6)
+    {
+        utilsInitTxSocketIPv6(aSocket, addr6, ifIndex);
+    }
+    else
+    {
+        fprintf(stderr, "No sock address for TX socket!\n");
+        exit(EXIT_FAILURE);
+    }
 
-    rval = setsockopt(fd, IPPROTO_IP, IP_ADD_MEMBERSHIP, &mreq, sizeof(mreq));
-    otEXPECT_ACTION(rval != -1, perror("setsockopt(RxFd, IP_ADD_MEMBERSHIP)"));
+    utilsInitRxSocket(aSocket, (addr6 ? NULL : addr4), ifIndex);
+    aSocket->mInitialized = true;
+    aSocket->mIPv6        = (addr6 != NULL);
 
-    sockaddr.sin_family      = AF_INET;
-    sockaddr.sin_port        = htons(aSocket->mPortBase);
-    sockaddr.sin_addr.s_addr = inet_addr(UTILS_SOCKET_GROUP_ADDR);
+exit:
+    freeifaddrs(ifaddr);
+    return aSocket->mInitialized;
+}
 
-    rval = bind(fd, (struct sockaddr *)&sockaddr, sizeof(sockaddr));
-    otEXPECT_ACTION(rval != -1, perror("bind(RxFd)"));
+static bool tryInitSocketIPv4(utilsSocket *aSocket, const char *aLocalHost)
+{
+    struct in_addr addr4;
 
-    aSocket->mRxFd = fd;
+    otEXPECT(inet_pton(AF_INET, aLocalHost, &addr4));
 
+    utilsInitTxSocketIPv4(aSocket, &addr4);
+    utilsInitRxSocket(aSocket, &addr4, 0);
     aSocket->mInitialized = true;
+    aSocket->mIPv6        = false;
+
+exit:
+    return aSocket->mInitialized;
+}
+
+static bool tryInitSocketIPv6(utilsSocket *aSocket, const char *aLocalHost)
+{
+    struct in6_addr addr6;
+    struct ifaddrs *ifaddr = NULL;
+
+    otEXPECT(inet_pton(AF_INET6, aLocalHost, &addr6));
+
+    if (getifaddrs(&ifaddr) == -1)
+    {
+        perror("getifaddrs");
+        exit(EXIT_FAILURE);
+    }
+
+    for (struct ifaddrs *ifa = ifaddr; ifa != NULL; ifa = ifa->ifa_next)
+    {
+        if (ifa->ifa_addr == NULL || ifa->ifa_addr->sa_family != AF_INET6)
+        {
+            continue;
+        }
+
+        const struct sockaddr_in6 *sockaddr6 = (const struct sockaddr_in6 *)ifa->ifa_addr;
+        if (memcmp(&sockaddr6->sin6_addr, &addr6, sizeof(addr6)))
+        {
+            continue;
+        }
+
+        unsigned int ifIndex = if_nametoindex(ifa->ifa_name);
+        if (ifIndex == 0)
+        {
+            perror("if_nametoindex");
+            exit(EXIT_FAILURE);
+        }
+
+        utilsInitTxSocketIPv6(aSocket, &addr6, ifIndex);
+        utilsInitRxSocket(aSocket, NULL, ifIndex);
+        aSocket->mInitialized = true;
+        aSocket->mIPv6        = true;
+        break;
+    }
+
+    freeifaddrs(ifaddr);
 
 exit:
-    if (!aSocket->mInitialized)
+    return aSocket->mInitialized;
+}
+
+void utilsInitSocket(utilsSocket *aSocket, uint16_t aPortBase)
+{
+    aSocket->mInitialized = false;
+    aSocket->mPortBase    = aPortBase;
+    aSocket->mTxFd        = -1;
+    aSocket->mRxFd        = -1;
+    aSocket->mPort        = (uint16_t)(aSocket->mPortBase + gNodeId);
+
+    if (!tryInitSocketIfname(aSocket, gLocalHost) && tryInitSocketIPv4(aSocket, gLocalHost) &&
+        tryInitSocketIPv6(aSocket, gLocalHost))
```

Comment:
```suggestion
    if (!tryInitSocketIfname(aSocket, gLocalHost) && !tryInitSocketIPv4(aSocket, gLocalHost) &&
        !tryInitSocketIPv6(aSocket, gLocalHost))
```

Should it be like this?